clients/src/main/java/org/apache/kafka/common/requests/WriteTxnMarkerResponse.java

```diff
@@ -0,0 +1,125 @@
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.protocol.types.Struct;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class WriteTxnMarkerResponse extends AbstractResponse {
+    private static final String TOPIC_PARTITIONS_KEY_NAME = "topic_partitions";
+    private static final String PARTITIONS_KEY_NAME = "partitions";
+    private static final String TOPIC_KEY_NAME = "topic";
+    private static final String PARTITION_KEY_NAME = "partition";
+    private static final String ERROR_CODE_KEY_NAME = "error_code";
+
+    // Possible error codes:
+    //   CorruptRecord
+    //   InvalidProducerEpoch
+    //   UnknownTopicOrPartition
+    //   NotLeaderForPartition
+    //   MessageTooLarge
+    //   RecordListTooLarge
+    //   NotEnoughReplicas
+    //   NotEnoughReplicasAfterAppend
+    //   InvalidRequiredAcks
+
+    private final Map<TopicPartition, Errors> errors;
+
+    public WriteTxnMarkerResponse(Map<TopicPartition, Errors> errors) {
+        this.errors = errors;
+    }
+
+    public WriteTxnMarkerResponse(Struct struct) {
+        Map<TopicPartition, Errors> errors = new HashMap<>();
+        Object[] topicPartitionsArray = struct.getArray(TOPIC_PARTITIONS_KEY_NAME);
+        for (Object topicPartitionObj : topicPartitionsArray) {
+            Struct topicPartitionStruct = (Struct) topicPartitionObj;
+            String topic = topicPartitionStruct.getString(TOPIC_KEY_NAME);
+            for (Object partitionObj : topicPartitionStruct.getArray(PARTITIONS_KEY_NAME)) {
+                Struct partitionStruct = (Struct) partitionObj;
+                Integer partition = partitionStruct.getInt(PARTITION_KEY_NAME);
+                Errors error = Errors.forCode(partitionStruct.getShort(ERROR_CODE_KEY_NAME));
+                errors.put(new TopicPartition(topic, partition), error);
+            }
+        }
+        this.errors = errors;
+    }
+
+    @Override
+    protected Struct toStruct(short version) {
+        Struct struct = new Struct(ApiKeys.WRITE_TXN_MARKER.responseSchema(version));
+        Map<String, List<PartitionAndError>> mappedPartitions = new HashMap<>();
+        for (Map.Entry<TopicPartition, Errors> partitionErrorEntry : errors.entrySet()) {
+            TopicPartition partition = partitionErrorEntry.getKey();
+            List<PartitionAndError> partitionList = mappedPartitions.get(partition.topic());
+            if (partitionList == null) {
+                partitionList = new ArrayList<>();
+                mappedPartitions.put(partition.topic(), partitionList);
+            }
+            partitionList.add(new PartitionAndError(partition.partition(), partitionErrorEntry.getValue()));
+        }
+
+        Object[] partitionsArray = new Object[mappedPartitions.size()];
+        int i = 0;
```

Comment:
Since we may likely deprecate Java 7 in 0.11.0.0: could we use `Collector#groupBy` then?